driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/BatchCursorPublisherErrorTest.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client;
+
+import com.mongodb.client.model.Aggregates;
+import com.mongodb.client.model.Filters;
+import org.bson.Document;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.DynamicTest;
+import org.junit.jupiter.api.TestFactory;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.mongodb.ClusterFixture.TIMEOUT_DURATION;
+import static com.mongodb.reactivestreams.client.Fixture.drop;
+import static com.mongodb.reactivestreams.client.Fixture.getDefaultDatabase;
+import static com.mongodb.reactivestreams.client.Fixture.getMongoClient;
+import static com.mongodb.reactivestreams.client.Fixture.serverVersionAtLeast;
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+import static java.util.stream.IntStream.rangeClosed;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+import static org.junit.jupiter.api.DynamicTest.dynamicTest;
+
+
+public class BatchCursorPublisherErrorTest {
+
+    private MongoCollection<Document> collection;
+
+    @BeforeEach
+    public void setup() {
+        assumeTrue(serverVersionAtLeast(3, 6));
+        collection = getDefaultDatabase().getCollection("changeStreamsCancellationTest");
+        Mono.from(collection.insertMany(rangeClosed(1, 11)
+                .boxed()
+                .map(i -> Document.parse(format("{a: %s}", i)))
+                .collect(Collectors.toList()))
+        ).block(TIMEOUT_DURATION);
+    }
+
+    @AfterEach
+    public void tearDown() {
+        if (collection != null) {
+            drop(collection.getNamespace());
+        }
+    }
+
+    @TestFactory
+    @DisplayName("test batch cursors close the cursor if onNext throws an error")
+    List<DynamicTest> testBatchCursorThrowsAnError() {
```

Comment:
I tried to see if this test actually catches the problem reported in JAVA-3976, and it does not seem like it does: the test passes on both 0b9292e530013c75a53d36d76fb5018a4dd257cf (immediately before [Fix reactive streams batch cursor regression. (#652)](4af9e9f2a541b8a89eadd7f91124b324015866eb)) and on [Fix reactive streams batch cursor regression. (#652)](4af9e9f2a541b8a89eadd7f91124b324015866eb).

However, if we add `System.out.println("AsyncBatchCursor close");` to `AsyncBatchCursor.close` and enable `DEBUG` logging in `logback-test.xml`, we can see that
- with the version immediately before the fix, `AsyncBatchCursor.close` is not called and `killCursor` commands are not issued (nevertheless, the test passes)
- with the fixed version, `AsyncBatchCursor.close` is called and we can see `killCursor` commands in the logs.

I am not sure what would be the right approach to test this. Maybe using the same technique the specification tests do: monitor the commands sent by the client to verify that `killCursors` commands are sent when expected?

Reply:
KillCursors won't be sent for all scenarios.

However, this tests onDispose / onCancel handle the error scenario, removing those calls, makes this test fail as the session is not returned.

Reply:
I understand that we may not need to always issue `killCursor`, but it seems reasonable to expect that we need to always close a `Closeable` resource (`AsyncBatchCursor`). While we do this as of [Fix reactive streams batch cursor regression. (#652)](https://github.com/mongodb/mongo-java-driver/pull/4af9e9f2a541b8a89eadd7f91124b324015866eb), this test does not seem to verify this behavior, and the lack of this behavior is a part of what was reported in JAVA-3976.

On the other hand, spending too much time trying to come up with a test for the aforementioned internal behavior may not be a good use of time. It's good to have this new test as is. LGTM.